File: solution/094BinaryTreeInorderTraversal/094BinaryTreeInorderTraversal.go
package main

import "fmt"

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversalRecursive(root *TreeNode, result []int) []int {
	if root != nil {
		result = inorderTraversalRecursive(root.Left, result)
		result = append(result, root.Val)
		result = inorderTraversalRecursive(root.Right, result)
	}
	return result
}

type stack []*TreeNode

func (s stack) Empty() bool       { return len(s) == 0 }
func (s stack) Peek() *TreeNode   { return s[len(s)-1] }
func (s *stack) Push(i *TreeNode) { (*s) = append((*s), i) }
func (s *stack) Pop() *TreeNode {
	d := (*s)[len(*s)-1]
	(*s) = (*s)[:len(*s)-1]
	return d
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	//result = inorderTraversalRecursive(root, result)
	var s stack
	m := make(map[*TreeNode]bool)
	if root != nil {
		s.Push(root)
	}
	for !s.Empty() {
		p := s.Peek()
		ok := p.Left != nil
		if ok && !m[p.Left] {
			s.Push(p.Left)
		} else {
			r := s.Pop()
			m[r] = true
			result = append(result, r.Val)
			if r.Right != nil {
				s.Push(r.Right)
			}
		}
	}
	return result
}

func main() {
	p3 := TreeNode{3, nil, nil}
	p2 := TreeNode{2, &p3, nil}
	root := TreeNode{1, nil, &p2}
	fmt.Printf("%v\n", inorderTraversal(&root))
}

File: src/094BinaryTreeInorderTraversal/094BinaryTreeInorderTraversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution1 {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = vec![];
        let mut stack = vec![];
        let mut root = root;
        while root.is_some() || stack.len() > 0 {
            if let Some(node) = root {
                root = node.borrow_mut().left.take();
                stack.push(node);
            }else{
                let node = stack.pop().unwrap();
                result.push(node.borrow().val);
                root = node.borrow_mut().right.take();
            }
        }
        
        result
    }
}

impl Solution2 {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut values = vec![];
        Solution::inorder_traversal_helper(&root, &mut values);
        values
    }
    fn inorder_traversal_helper(root: &Option<Rc<RefCell<TreeNode>>>, values: &mut Vec<i32>)  {
        if let Some(node) = root {
            Solution::inorder_traversal_helper(&node.borrow().left, values);
            values.push(node.borrow().val);
            Solution::inorder_traversal_helper(&node.borrow().right, values);
        }
    }
}
